src/cascadia/Remoting/Monarch.cpp

```diff
@@ -130,63 +139,239 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
-        // in practice. The WindowManager should set the MRU peasant
-        // immediately as soon as it creates the monarch/peasant for the
-        // first window.
-        if (_peasants.size() > 0)
+        // Here, there's at least one MRU peasant.
+        //
+        // We're going to iterate over these peasants toll we find one that both:
+        // 1. Is alive
+        // 2. Meets our selection criteria (do we care if it is on this desktop?)
+        //
+        // If the peasant is dead, then we'll remove it, and try the next one.
+        // Once we find one that's alive, we'll either:
+        // * check if we only want a peasant on the current desktop, and if so,
+        //   check if this peasant is on the current desktop.
+        //   - If it isn't on the current desktop, we'll loop again, on the
+        //     following peasant.
+        // * If we don't care, then we'll just return that one.
+
+        int positionInList = 0;
+        while (_mruPeasants.begin() + positionInList < _mruPeasants.end())
         {
-            try
+            const auto mruWindowArgs{ *(_mruPeasants.begin() + positionInList) };
+            const auto peasant{ _getPeasant(mruWindowArgs.PeasantID()) };
+            if (!peasant)
             {
-                return _peasants.begin()->second.GetID();
+                TraceLoggingWrite(g_hRemotingProvider,
+                                  "Monarch_Collect_WasDead",
+                                  TraceLoggingUInt64(mruWindowArgs.PeasantID(),
+                                                     "peasantID",
+                                                     "We thought this peasant was the MRU one, but it was actually already dead."),
+                                  TraceLoggingGuid(mruWindowArgs.DesktopID(), "desktopGuid", "The GUID of the desktop the window is on"),
+                                  TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE));
+                // We'll go through the loop again. We removed the current one
+                // at positionInList, so the next one in positionInList will be
+                // a new, different peasant.
+                continue;
             }
-            catch (...)
+
+            if (limitToCurrentDesktop && _desktopManager)
             {
-                // This shouldn't really happen. If we're the monarch, then the
-                // first peasant should also _be us_. So we should be able to
-                // get our own ID.
-                return 0;
+                // Check if this peasant is actually on this desktop. We can't
+                // simply get the GUID of the current desktop. We have to ask if
+                // the HWND is on the current desktop.
+                BOOL onCurrentDesktop{ false };
+
+                // SUCCEEDED_LOG will log if it failed, and return true if it SUCCEEDED
+                if (SUCCEEDED_LOG(_desktopManager->IsWindowOnCurrentVirtualDesktop((HWND)mruWindowArgs.Hwnd(),

```

Comment:
prefer a more expressive cast here, like `static_cast<HWND>`

```suggestion
                if (SUCCEEDED_LOG(_desktopManager->IsWindowOnCurrentVirtualDesktop(static_cast<HWND>(mruWindowArgs.Hwnd()),
```

Reply:
Oh, it does _not_ like that, I think because an HWND is technically a `void*`, so this would need to be a `reinterpret_cast`